File: frontend/js/views/real_return.js
// Real return — the Fisher inflation-adjusted return (exact + shortcut),
// after-tax real return, and purchasing power over time, via
// /calc/real-return. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const FIELDS = [
    ['nominal_return_pct', 'Nominal return (%)', 7],
    ['inflation_pct', 'Inflation (%)', 3],
    ['tax_rate_pct', 'Tax on return (%)', 15],
    ['principal_usd', 'Principal ($, optional)', 100000],
    ['years', 'Years (optional)', 20],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const pct = (n) => Number(n).toFixed(3) + '%';

export async function renderRealReturn(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rr.h1.title">// REAL (INFLATION-ADJUSTED) RETURN</span></h1>
        <p class="muted small" data-i18n="view.rr.hint.intro">
            A nominal return overstates your gain — inflation erodes purchasing power and taxes
            take a cut. The Fisher equation gives the exact real return: (1 + nominal) ÷ (1 +
            inflation) − 1. The "nominal − inflation" shortcut runs a touch high. After-tax real
            applies the tax to the return first. Over a horizon, the real future value is the
            principal's purchasing power in today's dollars. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.rr.h2.inputs">The numbers</h2>
            <form id="rr-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.rr.label.${key}">${label}</span>
                        <input type="number" step="0.01" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
        </div>
        <div id="rr-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#rr-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const r = await api.calcRealReturn(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.rr.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#rr-result');
    const realCls = r.real_return_pct >= 0 ? 'pos' : 'neg';
    const atrCls = r.after_tax_real_pct >= 0 ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.rr.h2.result">Real return</h2>
            <div class="cards">
                <div class="card ${realCls}"><div class="label" data-i18n="view.rr.card.real">Real return (Fisher)</div>
                    <div class="value ${realCls}">${pct(r.real_return_pct)}</div></div>
                <div class="card ${atrCls}"><div class="label" data-i18n="view.rr.card.aftertax">After-tax real</div>
                    <div class="value ${atrCls}">${pct(r.after_tax_real_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rr.card.fv">Real future value</div>
                    <div class="value">${money(r.real_future_value_usd)}</div></div>
            </div>
            <table class="data-table">
                <thead><tr><th data-i18n="view.rr.col.line">Line</th><th data-i18n="view.rr.col.rate">Rate</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.rr.row.exact">Real (exact)</td><td>${pct(r.real_return_pct)}</td></tr>
                    <tr><td data-i18n="view.rr.row.approx">Real (shortcut)</td><td>${pct(r.approx_real_return_pct)}</td></tr>
                    <tr><td data-i18n="view.rr.row.atnom">After-tax nominal</td><td>${pct(r.after_tax_nominal_pct)}</td></tr>
                    <tr class="emph"><td data-i18n="view.rr.row.atreal">After-tax real</td><td class="${atrCls}">${pct(r.after_tax_real_pct)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/real_return.rs
//! Real (inflation-adjusted) return — what your money actually buys.
//!
//! A nominal return overstates your gain: inflation erodes purchasing power
//! and taxes take a cut. The Fisher equation gives the exact real return:
//!
//!   * real = (1 + nominal) / (1 + inflation) − 1   (exact)
//!   * real ≈ nominal − inflation                    (the common shortcut)
//!   * after-tax real = (1 + nominal×(1−tax)) / (1 + inflation) − 1
//!
//! The shortcut overstates the real return slightly; the gap grows with the
//! rates. Over a horizon, the real future value shows the principal's
//! purchasing power in today's dollars. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RealReturnInput {
    pub nominal_return_pct: f64,
    pub inflation_pct: f64,
    /// Tax rate on the return (0 ⇒ pre-tax).
    #[serde(default)]
    pub tax_rate_pct: f64,
    /// Principal and horizon for the purchasing-power projection (optional).
    #[serde(default)]
    pub principal_usd: f64,
    #[serde(default)]
    pub years: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct RealReturnResult {
    /// Exact Fisher real return.
    pub real_return_pct: f64,
    /// nominal − inflation (the shortcut).
    pub approx_real_return_pct: f64,
    /// After-tax nominal return.
    pub after_tax_nominal_pct: f64,
    /// After-tax real return.
    pub after_tax_real_pct: f64,
    /// Principal grown at the real return for `years` (today's dollars).
    pub real_future_value_usd: f64,
}

pub fn analyze(i: &RealReturnInput) -> RealReturnResult {
    let n = i.nominal_return_pct / 100.0;
    let infl = i.inflation_pct / 100.0;
    let tax = i.tax_rate_pct / 100.0;

    let real = (1.0 + n) / (1.0 + infl) - 1.0;
    let after_tax_nominal = n * (1.0 - tax);
    let after_tax_real = (1.0 + after_tax_nominal) / (1.0 + infl) - 1.0;

    let real_fv = i.principal_usd * (1.0 + real).powf(i.years.max(0.0));

    RealReturnResult {
        real_return_pct: real * 100.0,
        approx_real_return_pct: i.nominal_return_pct - i.inflation_pct,
        after_tax_nominal_pct: after_tax_nominal * 100.0,
        after_tax_real_pct: after_tax_real * 100.0,
        real_future_value_usd: real_fv,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn inp(nom: f64, infl: f64) -> RealReturnInput {
        RealReturnInput {
            nominal_return_pct: nom,
            inflation_pct: infl,
            tax_rate_pct: 0.0,
            principal_usd: 0.0,
            years: 0.0,
        }
    }

    #[test]
    fn fisher_exact_real_return() {
        // (1.07/1.03) − 1 = 3.8835%.
        let r = analyze(&inp(7.0, 3.0));
        assert!((r.real_return_pct - ((1.07 / 1.03 - 1.0) * 100.0)).abs() < 1e-9);
        assert!((r.real_return_pct - 3.8835).abs() < 1e-3);
    }

    #[test]
    fn approx_is_simple_difference() {
        let r = analyze(&inp(7.0, 3.0));
        assert!((r.approx_real_return_pct - 4.0).abs() < 1e-9);
    }

    #[test]
    fn approx_overstates_exact() {
        let r = analyze(&inp(7.0, 3.0));
        assert!(r.approx_real_return_pct > r.real_return_pct);
    }

    #[test]
    fn after_tax_nominal_applies_tax() {
        let r = analyze(&RealReturnInput { tax_rate_pct: 15.0, ..inp(7.0, 3.0) });
        assert!((r.after_tax_nominal_pct - 7.0 * 0.85).abs() < 1e-9);
    }

    #[test]
    fn after_tax_real_below_pretax_real() {
        let r = analyze(&RealReturnInput { tax_rate_pct: 15.0, ..inp(7.0, 3.0) });
        assert!(r.after_tax_real_pct < r.real_return_pct);
    }

    #[test]
    fn zero_inflation_real_equals_nominal() {
        let r = analyze(&inp(7.0, 0.0));
        assert!((r.real_return_pct - 7.0).abs() < 1e-9);
    }

    #[test]
    fn negative_real_when_inflation_exceeds_nominal() {
        // 2% nominal, 5% inflation → losing purchasing power.
        let r = analyze(&inp(2.0, 5.0));
        assert!(r.real_return_pct < 0.0);
    }

    #[test]
    fn real_future_value_uses_real_return() {
        let r = analyze(&RealReturnInput { principal_usd: 100_000.0, years: 20.0, ..inp(7.0, 3.0) });
        let real: f64 = 1.07 / 1.03 - 1.0;
        assert!((r.real_future_value_usd - 100_000.0 * (1.0 + real).powi(20)).abs() < 1e-2);
    }
}
